java/ql/src/semmle/code/java/dataflow/FlowSources.qll

```diff
@@ -304,17 +305,22 @@ class ReverseDNSMethod extends Method {
   }
 }
 
-/** Android `Intent` that may have come from a hostile application. */
-class AndroidIntentInput extends DataFlow::Node {
+/** Exported Android `Intent` that may have come from a hostile application. */
+class AndroidIntentInput extends RemoteFlowSource {
   AndroidIntentInput() {
-    exists(MethodAccess ma, AndroidGetIntentMethod m |
-      ma.getMethod().overrides*(m) and
-      this.asExpr() = ma
-    )
-    or
-    exists(Method m, AndroidReceiveIntentMethod rI |
-      m.overrides*(rI) and
-      this.asParameter() = m.getParameter(1)
+    this.getEnclosingCallable().getDeclaringType().(AndroidComponent).isExported() and
```

Comment:
`getEnclosingCallable` is right for the ReceiveIntentMethod case, but not for GetIntent. For example, consider

```java
class OtherClass {

	public void test(IntentSources is) {
		String trouble = is.getIntent().getStringExtra("key");
		Runtime.getRuntime().exec(trouble);
	}

}
```

In this case it is not the enclosing callable but the method receiver that has the desired type.

Consider therefore something like

```
    exists(AndroidComponent exportedType |
      exportedType.isExported() |
      exists(MethodAccess ma, AndroidGetIntentMethod m |
        ma.getMethod().overrides*(m) and
        this.asExpr() = ma and
        exportedType = ma.getQualifier().getType()
      )
      or
      exists(Method m, AndroidReceiveIntentMethod rI |
        m.overrides*(rI) and
        this.asParameter() = m.getParameter(1) and
        exportedType = m.getDeclaringType()
      )
    )
```

Reply:
Thanks @smowton for the sample code. However, with 
```
exportedType = ma.getQualifier().getType()
```
a test program like the following won't return correct result:
```
public class UnsafeAndroidActivity extends Activity {
	public void onCreate(Bundle savedInstanceState) {
		Intent intent = getIntent();
		String thisUrl = intent.getStringExtra("url");
		WebView wv = (WebView) findViewById(-1);
		wv.loadUrl(thisUrl);
	}
}
```

Typically Android programs pass intents between them instead of Android components like <code>IntentSources is</code> as a method parameter directly. 

Did I miss something?

Thanks,
@luchua-bc

Reply:
Oops, instead of `getQualifier().getType()` I should have used `getReceiverType()`, whose definition combines our two approaches:

```codeql
  RefType getReceiverType() {
    result = getQualifier().getType()
    or
    not hasQualifier() and result = getEnclosingCallable().getDeclaringType()
  }
```

Reply:
Instead of the above, I would recommend `ma.getMethod().getDeclaringType()`. That's much more direct and doesn't depend on whether there's a qualifier or not. The `getReceiverType()` unfortunately isn't entirely accurate for e.g. nested types.

Reply:
Thanks @smowton. Changing to <code>exportedType = ma.getReceiverType()</code> helps.

Reply:
Thanks @aschackmull. I will stay with <code>ma.getMethod().getDeclaringType()</code> since that covers the typical Android usage.